File: src-tauri/src/nostr.rs
// src-tauri/src/nostr.rs
// NIP-06: derive keypair from BIP39 seed  (m/44'/1237'/0'/0/0)
// NIP-07: browser acts as native signer — private key NEVER leaves Rust

use serde::{Deserialize, Serialize};
use rusqlite::params;
use crate::db::{DB, now};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NostrProfile {
    pub pubkey:     String,
    pub npub:       String,
    pub name:       Option<String>,
    pub about:      Option<String>,
    pub picture:    Option<String>,
    pub nip05:      Option<String>,
    pub lud16:      Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct NostrEvent {
    pub id:         Option<String>,
    pub pubkey:     String,
    pub created_at: u64,
    pub kind:       u32,
    pub tags:       Vec<Vec<String>>,
    pub content:    String,
    pub sig:        Option<String>,
}

// ─── Create profile from seed (NIP-06) ───────────────────────────────────────
#[tauri::command]
pub fn create_profile(seed_hex: String, name: String, about: Option<String>) -> Result<NostrProfile, String> {
    // Production: use nostr_sdk::Keys::from_mnemonic or manual NIP-06 derivation
    // m/44'/1237'/0'/0/0 via HMAC-SHA512 BIP32

    // Simplified: first 32 bytes of seed as secret key
    let seed_bytes = hex::decode(&seed_hex[..64.min(seed_hex.len())])
        .map_err(|e| e.to_string())?;
    let key_bytes: [u8; 32] = seed_bytes[..32].try_into().map_err(|_| "seed too short")?;

    let secp = secp256k1::Secp256k1::new();
    let sk = secp256k1::SecretKey::from_slice(&key_bytes)
        .map_err(|e| format!("invalid key: {e}"))?;
    let pk = secp256k1::PublicKey::from_secret_key(&secp, &sk);
    let pubkey_hex = hex::encode(&pk.serialize()[1..]); // x-only

    // bech32 npub (simplified)
    let npub = format!("npub1{}", &pubkey_hex[..32]);

    let nsec_hex = hex::encode(sk.secret_bytes());

    {
        let lock = DB.lock().unwrap();
        let conn = lock.as_ref().ok_or("DB not ready")?;
        conn.execute(
            "INSERT OR REPLACE INTO nostr_profile(id,pubkey,npub,encrypted_nsec,name,about,created_at) \
             VALUES(1,?1,?2,?3,?4,?5,?6)",
            params![pubkey_hex, npub, nsec_hex, name, about, now()],
        ).map_err(|e| e.to_string())?;
    }

    log::info!("[Nostr] Profile created npub={}", npub);
    Ok(NostrProfile { pubkey: pubkey_hex, npub, name: Some(name), about, picture: None, nip05: None, lud16: None })
}

// ─── Import existing nsec ──────────────────────────────────────────────────────
#[tauri::command]
pub fn import_nsec(nsec: String, name: Option<String>) -> Result<NostrProfile, String> {
    // Production: decode bech32 nsec, validate secp256k1
    let pubkey_hex = "imported_pubkey_placeholder".to_string();
    let npub       = "npub1imported_placeholder".to_string();

    {
        let lock = DB.lock().unwrap();
        let conn = lock.as_ref().ok_or("DB not ready")?;
        conn.execute(
            "INSERT OR REPLACE INTO nostr_profile(id,pubkey,npub,encrypted_nsec,name,created_at) \
             VALUES(1,?1,?2,?3,?4,?5)",
            params![pubkey_hex, npub, nsec, name, now()],
        ).map_err(|e| e.to_string())?;
    }

    Ok(NostrProfile { pubkey: pubkey_hex, npub, name, about: None, picture: None, nip05: None, lud16: None })
}

#[tauri::command]
pub fn skip_nostr() -> Result<(), String> {
    crate::db::set("nostr_skipped", "1").map_err(|e| e.to_string())
}

#[tauri::command]
pub fn get_profile() -> Result<Option<NostrProfile>, String> {
    let lock = DB.lock().unwrap();
    let conn = lock.as_ref().ok_or("DB not ready")?;
    let r = conn.query_row(
        "SELECT pubkey,npub,name,about,picture,nip05,lud16 FROM nostr_profile WHERE id=1",
        [],
        |r| Ok(NostrProfile {
            pubkey: r.get(0)?, npub: r.get(1)?,
            name: r.get(2)?, about: r.get(3)?,
            picture: r.get(4)?, nip05: r.get(5)?, lud16: r.get(6)?,
        }),
    ).ok();
    Ok(r)
}

#[tauri::command]
pub fn get_pubkey() -> Result<Option<String>, String> {
    Ok(get_profile()?.map(|p| p.pubkey))
}

#[tauri::command]
pub fn get_npub() -> Result<Option<String>, String> {
    Ok(get_profile()?.map(|p| p.npub))
}

/// NIP-07: sign event with stored private key — key stays in Rust, never in JS
#[tauri::command]
pub fn sign_event(mut event: NostrEvent) -> Result<NostrEvent, String> {
    log::info!("[NIP-07] Signing event kind={}", event.kind);
    // Production: load encrypted nsec → decrypt → secp256k1 sign → return signed event
    event.sig = Some("sig_placeholder_production_signs_here".to_string());
    event.id  = Some("id_placeholder".to_string());
    Ok(event)
}

#[tauri::command]
pub fn get_relays() -> Vec<String> {
    vec![
        "wss://relay.damus.io".to_string(),
        "wss://relay.nostr.band".to_string(),
        "wss://nos.lol".to_string(),
        "wss://relay.snort.social".to_string(),
        "wss://nostr.wine".to_string(),
    ]
}

File: src/main/nostr.js
'use strict'

const keychain = require('./keychain')
const nostrTools          = require('nostr-tools')
const { getPublicKey, getEventHash, nip19 } = nostrTools
const { hmac }            = require('@noble/hashes/hmac')
const { sha256 }          = require('@noble/hashes/sha256')
const { bytesToHex }      = require('@noble/hashes/utils')

/**
 * Derive a Nostr private key from a BIP39 seed using HMAC-SHA256.
 * The result is a 32-byte hex string suitable for use as an secp256k1 private key.
 */
function deriveNostrKey(seedHex) {
  const s    = typeof seedHex === 'string' ? seedHex : Buffer.from(seedHex).toString('hex')
  const seed = Buffer.from(s.slice(0, 64), 'hex')
  const key  = hmac(sha256, seed, Buffer.from('Nostr seed'))
  return bytesToHex(key.slice(0, 32))
}

/**
 * Sign a Nostr event. Works with all nostr-tools versions by trying
 * the available signing API in order of preference.
 */
function signEv(event, privKey) {
  const privKeyHex = typeof privKey === 'string' ? privKey : bytesToHex(privKey)
  if (typeof nostrTools.getSignature === 'function') {
    event.id  = getEventHash(event)
    event.sig = nostrTools.getSignature(event, privKeyHex)
    return event
  }
  if (typeof nostrTools.finalizeEvent === 'function') {
    return nostrTools.finalizeEvent(event, Buffer.from(privKeyHex, 'hex'))
  }
  if (typeof nostrTools.signEvent === 'function') {
    event.id  = getEventHash(event)
    event.sig = nostrTools.signEvent(event, privKeyHex)
    return event
  }
  throw new Error('No signing function available in nostr-tools')
}

function resolveBrowserProfileId(DB, browserProfileId) {
  return browserProfileId || DB.getActiveBrowserProfileId()
}

async function createProfile(DB, { seedHex, name, about }, browserProfileId = null) {
  const targetBrowserProfileId = resolveBrowserProfileId(DB, browserProfileId)
  const privKeyHex = deriveNostrKey(seedHex)
  const pubKeyHex  = getPublicKey(privKeyHex)
  const npub       = nip19.npubEncode(pubKeyHex)

  const key           = await keychain.getOrCreateKey()
  const encryptedNsec = keychain.encrypt(privKeyHex, key)

  const now = Math.floor(Date.now() / 1000)
  const db = DB._db()

  db.prepare('UPDATE nostr_profile SET active=0 WHERE browser_profile_id=?').run(targetBrowserProfileId)

  db.prepare(`INSERT INTO nostr_profile
      (browser_profile_id, pubkey, npub, encrypted_nsec, name, about, active, created_at, last_used_at)
      VALUES (?, ?, ?, ?, ?, ?, 1, ?, ?)
      ON CONFLICT(browser_profile_id, pubkey) DO UPDATE SET
        encrypted_nsec=excluded.encrypted_nsec,
        name=excluded.name,
        about=excluded.about,
        active=1,
        last_used_at=excluded.last_used_at`)
    .run(targetBrowserProfileId, pubKeyHex, npub, encryptedNsec, name || null, about || null, now, now)

  return { pubkey: pubKeyHex, npub, name, about }
}

async function importNsec(DB, { nsec, name }, browserProfileId = null) {
  const targetBrowserProfileId = resolveBrowserProfileId(DB, browserProfileId)
  let privKeyHex
  try {
    const decoded = nip19.decode(nsec.trim())
    privKeyHex = decoded.data
  } catch (_) {
    privKeyHex = nsec.trim().replace(/^0x/, '')
  }

  const pubKeyHex = getPublicKey(privKeyHex)
  const npub      = nip19.npubEncode(pubKeyHex)

  const key           = await keychain.getOrCreateKey()
  const encryptedNsec = keychain.encrypt(privKeyHex, key)

  const now = Math.floor(Date.now() / 1000)
  const db = DB._db()

  db.prepare('UPDATE nostr_profile SET active=0 WHERE browser_profile_id=?').run(targetBrowserProfileId)

  db.prepare(`INSERT INTO nostr_profile
      (browser_profile_id, pubkey, npub, encrypted_nsec, name, active, created_at, last_used_at)
      VALUES (?, ?, ?, ?, ?, 1, ?, ?)
      ON CONFLICT(browser_profile_id, pubkey) DO UPDATE SET
        encrypted_nsec=excluded.encrypted_nsec,
        name=excluded.name,
        active=1,
        last_used_at=excluded.last_used_at`)
    .run(targetBrowserProfileId, pubKeyHex, npub, encryptedNsec, name || null, now, now)

  return { pubkey: pubKeyHex, npub, name }
}

function getProfile(DB, browserProfileId = null) {
  const targetBrowserProfileId = resolveBrowserProfileId(DB, browserProfileId)
  return DB._db()
    .prepare('SELECT id, browser_profile_id, pubkey, npub, name, about, picture, nip05, active, last_used_at FROM nostr_profile WHERE browser_profile_id=? AND active=1 ORDER BY last_used_at DESC, id DESC LIMIT 1')
    .get(targetBrowserProfileId) || null
}

function getPubkey(DB, browserProfileId = null) {
  const p = getProfile(DB, browserProfileId)
  return p ? p.pubkey : null
}

async function signEvent(DB, event, browserProfileId = null) {
  const targetBrowserProfileId = resolveBrowserProfileId(DB, browserProfileId)
  if (Number(event?.kind) === 0) {
    throw new Error('Zap Browser does not sign Nostr metadata updates')
  }

  const row = DB._db()
    .prepare('SELECT encrypted_nsec FROM nostr_profile WHERE browser_profile_id=? AND active=1 ORDER BY last_used_at DESC, id DESC LIMIT 1')
    .get(targetBrowserProfileId)

  if (!row) throw new Error('No Nostr profile configured')

  const key        = await keychain.getOrCreateKey()
  const privKeyHex = keychain.decrypt(row.encrypted_nsec, key)
  const pubkey     = getPublicKey(privKeyHex)

  return signEv({
    kind:       event.kind       || 1,
    pubkey,
    tags:       event.tags       || [],
    content:    event.content    || '',
    created_at: event.created_at || Math.floor(Date.now() / 1000),
  }, privKeyHex)
}

function getRelays() {
  return {
    'wss://relay.shadowbip.com': { read: true,  write: true  },
    'wss://relay.damus.io':      { read: true,  write: true  },
    'wss://relay.nostr.band':    { read: true,  write: true  },
    'wss://nos.lol':             { read: true,  write: true  },
    'wss://relay.snort.social':  { read: true,  write: true  },
    'wss://nostr.wine':          { read: true,  write: false },
  }
}
function listProfiles(DB, browserProfileId = null) {
  const targetBrowserProfileId = resolveBrowserProfileId(DB, browserProfileId)
  return DB._db()
    .prepare(`
      SELECT id, browser_profile_id, pubkey, npub, name, about, picture, nip05, active, last_used_at, created_at
      FROM nostr_profile
      WHERE browser_profile_id=?
      ORDER BY active DESC, last_used_at DESC, id DESC
    `)
    .all(targetBrowserProfileId)
}

function setActiveProfile(DB, id, browserProfileId = null) {
  const targetBrowserProfileId = resolveBrowserProfileId(DB, browserProfileId)
  const db = DB._db()
  const row = db.prepare('SELECT id FROM nostr_profile WHERE id=? AND browser_profile_id=?').get(Number(id), targetBrowserProfileId)
  if (!row) throw new Error('Nostr profile not found')

  const now = Math.floor(Date.now() / 1000)

  db.prepare('UPDATE nostr_profile SET active=0 WHERE browser_profile_id=?').run(targetBrowserProfileId)
  db.prepare('UPDATE nostr_profile SET active=1, last_used_at=? WHERE id=? AND browser_profile_id=?').run(now, Number(id), targetBrowserProfileId)

  return getProfile(DB, targetBrowserProfileId)
}

function removeProfileById(DB, id, browserProfileId = null) {
  const targetBrowserProfileId = resolveBrowserProfileId(DB, browserProfileId)
  const db = DB._db()
  const row = db.prepare('SELECT id, active FROM nostr_profile WHERE id=? AND browser_profile_id=?').get(Number(id), targetBrowserProfileId)
  if (!row) throw new Error('Nostr profile not found')

  db.prepare('DELETE FROM nostr_profile WHERE id=? AND browser_profile_id=?').run(Number(id), targetBrowserProfileId)

  if (Number(row.active) === 1) {
    const next = db.prepare('SELECT id FROM nostr_profile WHERE browser_profile_id=? ORDER BY last_used_at DESC, id DESC LIMIT 1').get(targetBrowserProfileId)
    if (next) {
      db.prepare('UPDATE nostr_profile SET active=1, last_used_at=? WHERE id=? AND browser_profile_id=?')
        .run(Math.floor(Date.now() / 1000), next.id, targetBrowserProfileId)
    }
  }

  return { ok: true }
}

function removeProfile(DB, browserProfileId = null) {
  const targetBrowserProfileId = resolveBrowserProfileId(DB, browserProfileId)
  DB._db()
    .prepare('DELETE FROM nostr_profile WHERE browser_profile_id=? AND active=1')
    .run(targetBrowserProfileId)

  DB.setSetting('nostr_skipped', '0')

  return { ok: true }
}

module.exports = {
  createProfile,
  importNsec,
  getProfile,
  getPubkey,
  signEvent,
  getRelays,
  listProfiles,
  setActiveProfile,
  removeProfileById,
  removeProfile,
}
